tests/federation/transport/test_knocking.py

```diff
@@ -0,0 +1,289 @@
+from collections import OrderedDict
+from typing import Dict, List
+
+from mock import Mock
+
+from twisted.internet.defer import succeed
+
+from synapse import event_auth
+from synapse.api.constants import EventTypes
+from synapse.api.room_versions import RoomVersions
+from synapse.config.ratelimiting import FederationRateLimitConfig
+from synapse.events import builder
+from synapse.federation.transport import server as federation_server
+from synapse.rest import admin
+from synapse.rest.client.v1 import login, room
+from synapse.server import HomeServer
+from synapse.types import RoomAlias
+from synapse.util.ratelimitutils import FederationRateLimiter
+
+from tests.test_utils import event_injection, make_awaitable
+from tests.unittest import FederatingHomeserverTestCase, HomeserverTestCase
+
+# An identifier to use while MSC2304 is not in a stable release of the spec
+KNOCK_UNSTABLE_IDENTIFIER = "xyz.amorgan.knock"
+
+# An event type that we do not expect to be given to users knocking on a room
+SECRET_STATE_EVENT_TYPE = "com.example.secret"
+
+
+class FederationKnockingTestCase(FederatingHomeserverTestCase):
+    servlets = [
+        admin.register_servlets,
+        room.register_servlets,
+        login.register_servlets,
+    ]
+
+    def prepare(self, reactor, clock, homeserver):
+        self.store = homeserver.get_datastore()
+
+        # Have this homeserver auto-approve all event signature checking
+        def approve_all_signature_checking(_, ev):
+            return [succeed(ev[0])]
+
+        homeserver.get_federation_server()._check_sigs_and_hashes = (
+            approve_all_signature_checking
+        )
+
+        # Have this homeserver skip event auth checks.
+        #
+        # While this prevent membership transistion checks, that is already
+        # tested elsewhere
+        event_auth.check = Mock(return_value=make_awaitable(None))
+
+        class Authenticator:
```

Comment:
What is this for?

Reply:
Ah yes, this is for bypassing the need to provide authentication for federation requests. I'll add a comment.